### ISSUE TITLE
cv::compare() and cv::cuda::compare() might give different result

### ISSUE THREAD

**Issue by chacha21** [opencv/opencv#24129]
### System Information

OpenCV 4.7.0
Windows 7 64b
latest Visual Studio 2019

### Detailed description

When using a non-integer threshold with an integer matrix, the internal rounding of the threshold differs between cv::compare() and cv::cuda::compare(), leading to different results.

An easy workaround is to manually rounding the threshold to the desired edge (floor(), ceil(), round()...) before calling compare(), but it is less handy for the generic case where the threshold can be the result of an algorithm, and the input matrix has no specific depth.
I have tracked it down because it was the root of strange differences between CPU and (supposedly equivalent) GPU code.

Changing the rounding behaviour of `cv::compare()` or `cv::cuda::compare()` to make them match might be a breaking change for existing code.

Perhaps only extra documentation is needed here.

### Steps to reproduce

```
    cv::Mat m1_cpu(1, 1, CV_16U);
    cv::Mat m2_cpu(1, 1, CV_8U);
    cv::Mat m3_cpu(1, 1, CV_8U);
    cv::cuda::GpuMat m1_gpu(1, 1, CV_16U);
    cv::cuda::GpuMat m2_gpu(1, 1, CV_8U);
    m1_cpu = cv::Scalar::all(11);
    m1_gpu.upload(m1_cpu);
    cv::compare(m1_cpu, cv::Scalar::all(10.6), m2_cpu, cv::CMP_GT);
    cv::cuda::compare(m1_gpu, cv::Scalar::all(10.6), m2_gpu, cv::CMP_GT);
    m2_gpu.download(m3_cpu);
    const double n = cv::norm(m2_cpu, m3_cpu, cv::NORM_INF);
    CV_Assert(fabs(n)<1e-6);
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [ ] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [ ] I updated to the latest OpenCV version and the issue is still there
- [ ] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by cudawarped** [comment]
So the issue is that internally the CUDA routine only has overloads for comparing the same types in [cmp_scalar](https://github.com/opencv/opencv_contrib/blob/daaf645151b7afbafabfacf71ae2880cf6fc904e/modules/cudaarithm/src/cuda/cmp_scalar.cu#L169) with the conversion internally rounding upwards as
```
cv::Scalar_<ushort> value_(cv::Scalar(10.6)) (== 11)
```

> Perhaps only extra documentation is needed here.

Seems like a reasonable solution to me given that the underlying operation [arithm_op](https://github.com/opencv/opencv_contrib/blob/daaf645151b7afbafabfacf71ae2880cf6fc904e/modules/cudaarithm/src/element_operations.cpp#L296) is used everywhere and would be difficult to change.  

> and the input matrix has no specific depth.

The input matrix should always have a specific depth.  Couldn't this be used to round the scalar input on both host and device routines?

**Comment by chacha21** [comment]
> The input matrix should always have a specific depth

What I mean is like in the following pattern, where the depth does not have to be explicit in the written code
```

Mat src = getSomeMatFromContext();//no specific depth == I don't care here
Scalar s = getSomeAutoThresholdFromAlgorithm(src);//the depth is taken in account in delegate algorithms
Mat dst;
//I don't want to be compelled to write "if (!hasFloatingDepth(src)) s = Scalar(round(s[0]), round(s[1]), round(s[2]), round(s[3]));"
compare(src, s, dst, CMP_GT);

```

**Comment by cudawarped** [comment]
> //I don't want to be compelled to write "if (!hasFloatingDepth(src)) s = Scalar(round(s[0]), round(s[1]), round(s[2]), round(s[3]));"
> compare(src, s, dst, CMP_GT);

From your comment below I thought you were ok with that workaround if it was possible

> An easy workaround is to manually rounding the threshold to the desired edge (floor(), ceil(), round()...) before calling compare(), but it is less handy for the generic case where the threshold can be the result of an algorithm, and the input matrix has no specific depth.

The problem I see is that the a Scalar type cannot be intialized using Mat type and to get consistent results on the host and device you need both the `Mat` and the `Scalar` to have the same type.

I guess you could add a function which peformed this conversion, say `scalarWithMatType` and call that before your host conversion routine.
```
s = scalarWithMatType(s,src.type());
```

Alternatively you could constrain your src `Mat` to be a specific type and just use
```
s = Scalar_<constrainedType>(s);
```

@opencv-alalek is there an existing way to create a Scalar from a Mat type or a better alternative?

**Comment by chacha21** [comment]
> > //I don't want to be compelled to write "if (!hasFloatingDepth(src)) s = Scalar(round(s[0]), round(s[1]), round(s[2]), round(s[3]));"
> > compare(src, s, dst, CMP_GT);
> 
> From your comment below I thought you were ok with that workaround if it was possible

You understood correctly : it's not a big deal. I should have written "I would have preferred not to be able to, but I see not better solution".
I like the `scalarWithMatType()` idea which is a good hint of both the meaning, the goal, and the required underlying operation.

May I suggest a new Scalar member like `Scalar Scalar::clipForType(int type, enum direction)` with `direction` being somewhat similar to [MPFR_RNDN/MPFR_RNDD/MPFR_RNDU...](https://www.mpfr.org/mpfr-current/mpfr.html) ?

`compare(src, s.clipForType(src.type(), RNDN), dst, CMP_GT);`

Not perfect, because there would be the ambiguity of "rounding is ignored for a floating type", which is not obvious at first sight.

**Comment by cudawarped** [comment]
> May I suggest a new Scalar member like `Scalar Scalar::clipForType(int type, enum direction)` with `direction` being somewhat similar to [MPFR_RNDN/MPFR_RNDD/MPFR_RNDU...](https://www.mpfr.org/mpfr-current/mpfr.html) ?

I was thinking it should use saturate_cast to be consistent with the existing data type conversion

https://github.com/opencv/opencv/blob/f834736307c8328340aea48908484052170c9224/modules/core/include/opencv2/core/types.hpp#L2184

This "should" also be consistent with the way OpenCV performs a lot of its internal rounding.

**Comment by chacha21** [comment]
> I was thinking it should use `saturate_cast` to be consistent with the existing data type conversion
Not necessarily. For instance, if we perform `compare(mat8U, Scalar(256), CMP_GE)`, it would be wrong to adjust to `compare(mat8U, Scalar(255), CMP_GE)`.

**Comment by cudawarped** [comment]
> Not necessarily. For instance, if we perform `compare(mat8U, Scalar(256), CMP_GE)`, it would be wrong to adjust to `compare(mat8U, Scalar(255), CMP_GE)`.

True but for consistency between the host and device routines the result of
```
cv::compare(mat8U, Scalar::saturateCast(256,ma8U.type()), CMP_GE)
```
would be the same as
```
cuda::compare(GpuMat(mat8U), Scalar(256), CMP_GE)
```
becaue `Scalar::saturateCast(256,ma8U.type())` is equivalent to the operation performed internally inside [`cmpScalarImpl`](https://github.com/opencv/opencv_contrib/blob/daaf645151b7afbafabfacf71ae2880cf6fc904e/modules/cudaarithm/src/cuda/cmp_scalar.cu#L138)
```
  template <template <typename> class Op, typename T, int cn>
  void cmpScalarImpl(const GpuMat& src, cv::Scalar value, GpuMat& dst, Stream& stream)
  {
      typedef typename MakeVec<T, cn>::type src_type;
      typedef typename MakeVec<uchar, cn>::type dst_type;

      cv::Scalar_<T> value_ = value;
```
